pkg/concurrency/concurrency.go

```diff
@@ -0,0 +1,135 @@
+package concurrency
+
+import (
+	"sync"
+)
+
+// TODO: closeWhenDone bool parameter ::
+//			It currently is experimental, and therefore exists.
+//			Is there ever a situation to use false?
+
+// This function is used to merge the results of a slice of channels of a specific result type down to a single result channel of a second type.
+// mappingFn allows the caller to convert from the input type to the output type
+// if closeWhenDone is set to true, the output channel will be closed when all individual result channels of the slice have been closed - otherwise it will be left open for future use.
+// The same WaitGroup used to trigger that optional closing is returned for any other synchronization purposes.
+func SliceOfChannelsRawMerger[IndividualResultType any, OutputResultType any](individualResultChannels []<-chan IndividualResultType, outputChannel chan<- OutputResultType, mappingFn func(IndividualResultType) (OutputResultType, error), closeWhenDone bool) *sync.WaitGroup {

```

Comment:
My suggestion here is to structure the work payload into a single channel which carries a struct which contains channels. As channels are first-class citizien in go, it's much more clearer to have a channel that exchange a structure that holds channel for its results (e.g. tokenChannel, an error field, and an output channel). 

I think it gives much more control and makes the code much more easier to read overall. I don't think generics are really needed in this case at all.

So instead of managing 3 channels across all the code - which should consume these functions all over - we could have instead a struct which encompasses the business logic

Reply:
Do you mind giving me a quick example of what the "struct which contains channels" your envisioning here would look like? I think the idea sounds good to me, but I'm not totally clear what you mean. Would that be something where we basically take the different "channel streams" of per-token, per-completion and overall result and bundle them into a single struct? If so... that definitely made sense. I had an earlier draft where I tried something akin to that, but before I settled on the different "streams" it was a pain to maintain and I never brought it back once stable as I was already used to the "weird" 3-channel approach.

I'm not sure when I initially started this project I knew about select...case channel blocking reads being able to handle the errors being on a separate channel from the data... With that "constraint" lifted, I think I'm starting to get a better picture

Reply:
https://go.dev/tour/concurrency/5 <-- it was here the whole time... I just forgot about it 🥇

Reply:
I'd still like that example if you get a chance... but `pkg/concurrency` has been dropped from this PR since I'm expecting to rework the related code significantly.